### PR TITLE
Use fsspec in load to resolve more paths/URLs from storage backends

### LINKED ISSUE #3691: cloud_io's `load` should use fsspec for compatibility with more storage systems
## 🐛 Bug

There's an inconsistency in which paths Lightning supports with `load` vs what `fsspec` can support: https://github.com/PyTorchLightning/pytorch-lightning/blob/b4256df121fae1955c78895cf32e2e4c66b6603d/pytorch_lightning/utilities/cloud_io.py#L27

This code should use `fsspec` to try to open the file. If there isn't a valid filesystem registered, then fallback to calling torch.hub.load. cc @awaelchli 

### To Reproduce

Pass a URL supported by `fsspec` but not torch.load(...) and try calling `cloud_io`'s `load` with it

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?
Fixes #3691 

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/cloud_io.py [hunk 1]
```diff
@@ -25,9 +25,15 @@
 
 
 def load(path_or_url: str, map_location=None):
-    if urlparse(path_or_url).scheme == "" or Path(path_or_url).drive:  # no scheme or with a drive letter
-        return torch.load(path_or_url, map_location=map_location)
-    return torch.hub.load_state_dict_from_url(path_or_url, map_location=map_location)
+    try:
+        fs = get_filesystem(path_or_url)
+    except ValueError:
```
Comment (ananthsub) [inline pytorch_lightning/utilities/cloud_io.py:30]: exception is thrown here: https://github.com/intake/filesystem_spec/blob/master/fsspec/registry.py#L190-L191

**Comment by awaelchli** [comment]
@Borda for these tests to pass, we need to `pip install fsspec[http]` extra in CI. Can we do that? Shouldn't be a problem right?

File: pytorch_lightning/utilities/cloud_io.py [hunk 2]
```diff
@@ -25,9 +25,15 @@
...
+        # Raised when the filesystem is neither local nor registered.
+        return torch.hub.load_state_dict_from_url(
+            path_or_url, map_location=map_location
+        )
```
Comment (Borda) [inline pytorch_lightning/utilities/cloud_io.py:34]:
```suggestion
        return torch.hub.load_state_dict_from_url(path_or_url, map_location=map_location)
```

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
it shall be fine

> @Borda <https://github.com/Borda> for these tests to pass, we need to pip
> install fsspec[http] extra in CI. Can we do that? Shouldn't be a problem
> right?
>

**Comment by ananthsub** [comment]
@Borda @awaelchli `aiohttps` needs to be included in the dependencies for the tests to run

**Comment by williamFalcon** [comment]
@ananthsub so this adds a dep to PL?

**Comment by ananthsub** [comment]
@williamFalcon it should already be there from relying on fsspec. https://github.com/intake/filesystem_spec/blob/master/setup.py#L45

from @awaelchli 
> oh so it needs be installed with [http] as brackets
> before we didn't get that because http was redirected to torch.hub.load
> yep, so if I do pip install fsspec[http], i get the aiohttp
> I guess we need at least http for the CI test suite

**Comment by Borda** [comment]
> @ananthsub so this adds a dep to PL?

if it is just for testing we are fine...

**Comment by ananthsub** [comment]
@Borda any pointers on how to include `fsspec`'s http dependencies for the tests?

**Comment by Borda** [comment]
> @Borda any pointers on how to include `fsspec`'s http dependencies for the tests?

Let me see, the case is that as we install requirements/devel.txt which chain all and pip do not like listed one package twice...

**Comment by Borda** [comment]
> @Borda any pointers on how to include `fsspec`'s http dependencies for the tests?

it seems that you already added them manually:
https://github.com/intake/filesystem_spec/blob/974e9c9b172f39816d929fe51e604ff7829ce1d3/setup.py#L45

File: requirements/test.txt [hunk 3]
```diff
@@ -13,3 +13,5 @@ pre-commit>=1.0
 
 cloudpickle>=1.2
 nltk>=3.3
+requests>=2.24
+aiohttp>=3.6.2
```
Comment (awaelchli) [inline requirements/test.txt:17]:
```suggestion
fsspec[http]
```
wouldn't this work? we can let fsspec handle the requests/aiohttp versions

Comment (Borda) [reply requirements/test.txt]: that shall be the same... but lets see

Comment (Borda) [reply requirements/test.txt]: seems to be failing the same way...

Comment (awaelchli) [reply requirements/test.txt:17]: yes, I was not suggesting this to fix anything, but to let fsspec handle the dependencies for http

File: pytorch_lightning/utilities/cloud_io.py [hunk 4]
```diff
@@ -25,9 +25,13 @@
...
 
 def load(path_or_url: str, map_location=None):
-    if urlparse(path_or_url).scheme == "" or Path(path_or_url).drive:  # no scheme or with a drive letter
-        return torch.load(path_or_url, map_location=map_location)
-    return torch.hub.load_state_dict_from_url(path_or_url, map_location=map_location)
+    try:
+        fs = get_filesystem(path_or_url)
+    except ValueError:
+        # Raised when the filesystem is neither local nor registered.
+        return torch.hub.load_state_dict_from_url(path_or_url, map_location=map_location)
```
Comment (awaelchli) [inline pytorch_lightning/utilities/cloud_io.py:32]: @ananthsub noob question, can fsspec savely replace torch.hub?

**Comment by Borda** [comment]
@awaelchli I have tried https://github.com/PyTorchLightning/pytorch-lightning/pull/3692#discussion_r496780035 and it was the same, then I reverted the change...

File: pytorch_lightning/utilities/cloud_io.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/cloud_io.py:32]: I believe the tests are failing because now the torch.hub.load is ignored when fsspec handles http.

Comment (ananthsub) [reply pytorch_lightning/utilities/cloud_io.py:29]: @awaelchli  this is super hacky but idk how to get fsspec and the unit tests to work with torch.hub.load. i think fsspec should be a direct replacement

**Pull request merged**